File: Years/2024/Days/1-5/1/Code/Java/part1Solution.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

import java.io.FileReader;
import java.io.BufferedReader;
import java.io.FileNotFoundException;

public class Part1Solution {
    public static void main(String[] args) {
        String fileContent = readFile("input1.txt");
        
        Integer[][] matrix = getMatrix(fileContent);

        Arrays.sort(matrix[0]);
        Arrays.sort(matrix[1]);
        
        long difference = getDifference(matrix[0], matrix[1]);
        System.out.println(difference);
    }

    public static String readFile(String filePath){
        try{
            FileReader fileReader = new FileReader(filePath);
            BufferedReader buffer = new BufferedReader(fileReader);
            StringBuilder stringBuilder = new StringBuilder();

            String line;
            while ((line = buffer.readLine()) != null){
                stringBuilder.append(line + '\n');
            }

            buffer.close();
            return stringBuilder.toString();
        }catch (FileNotFoundException exception){
            System.out.println("provided file path cannot be founbd");
        }catch (Exception exception){
            System.out.println("Something whent horribly wrong!!!");
            System.out.println(exception);
        }

        //If an exception occured, then return nothing
        return "";
    }

    public static Integer[][] getMatrix(String fileContent){
        String[] lines = fileContent.split("\n");

        List<Integer> column1 = new ArrayList<Integer>();
        List<Integer> column2 = new ArrayList<Integer>();
        
        for (String line : lines){
            String[] values = line.split(" ");
            column1.add(Integer.parseInt(values[0]));
            column2.add(Integer.parseInt(values[values.length - 1]));
        }

        return new Integer[][] {column1.toArray(new Integer[column1.size()]), column2.toArray(new Integer[column2.size()])};
    }

    public static long getDifference(Integer[] column1, Integer[] column2){
        long difference = 0;
        for (int i = 0; i < column1.length; i++){
            difference += Math.abs(column1[i] - column2[i]);
        }

        return difference;
    }
}

File: Years/2024/Days/1-5/1/Code/Python/part1Solution.py
import sys, re


from fshandler import read_file, FileExtensions, get_logger

input1_logger = get_logger(__name__)

def to_tuple(line) -> tuple[int, int]:
    splitted_line = re.sub(' +', ' ', line).split(' ')
    return int(splitted_line[0]), int(splitted_line[1])

def solve_part1(input1: str) -> int:
    try:
        input1_lines: list[str] = read_file(input1, FileExtensions.TXT)

        table = [to_tuple(line) for line in input1_lines if line != '\n']

        column1 = [column1 for column1, _ in table]
        column2 = [column2 for _, column2 in table]

        column1.sort()
        column2.sort()

        table = list(zip(column1, column2))

        total_sum = sum([abs(column1 - column2) for column1, column2 in table])
        return total_sum
    except Exception as error:
        input1_logger.error(error)
        return 0
    
def main():
    solution = solve_part1(str(sys.argv[1]))
    print(solution)

if __name__ == "__main__":
    main()
